File: hymd/compute_bond_forces.f90
subroutine cbf(f, r, box, i, j, r0, k, energy, bond_pr)
    ! Compute two-particle bond forces and energy
    !
    ! Parameters
    ! ---------
    ! f : (N,D) numpy.ndarray
    !     Forces for N particles in D dimensions. Changed in place.
    ! r : (N,D) numpy.ndarray
    !     Positions for N particles in D dimensions.
    ! box : (D,) numpy.ndarray
    !     D-dimensional simulation box size.
    ! a : (M,) numpys.ndarray
    !     Index of particle 1 for M individual two-particle bonds.
    ! b : (M,) numpy.ndarray
    !     Index of particle 2 for M individual two-particle bonds.
    ! r0 : (M,) numpy.ndarray
    !     Equilibrium bond distance for M individual two-particle bonds.
    ! k : (M,) numpy.ndarray
    !     Bond strength for M individual two-particle bonds.
    !
    ! Returns
    ! -------
    ! energy : float
    !     Total energy of all two-particle bonds.
    ! bond_pr : (3,) numpy.ndarray
    !     Total bond pressure due all two-particle bonds.
    !
    implicit none

    real(4), dimension(:,:), intent(in out) :: f
    real(4), dimension(:,:), intent(in)     :: r
    real(8), dimension(:),   intent(in)     :: box
    integer, dimension(:),   intent(in)     :: i
    integer, dimension(:),   intent(in)     :: j
    real(8), dimension(:),   intent(in)     :: r0
    real(8), dimension(:),   intent(in)     :: k
    real(8),                 intent(out)    :: energy
    real(4), dimension(3),   intent(out)    :: bond_pr

    integer :: ind, ii, jj
    real(8) :: rij, rij_x, rij_y, rij_z
    real(8) :: df
    real(8) :: bx, by, bz

    energy = 0.0d00
    bond_pr = 0.0d00 !Set x, y, z components to 0
    f = 0.0d00 ! Set all array elements

    bx = 1.0d00 / box(1)
    by = 1.0d00 / box(2)
    bz = 1.0d00 / box(3)

    do ind = 1, size(i)
      ii = i(ind) + 1
      jj = j(ind) + 1

      rij_x = r(jj, 1) - r(ii, 1)
      rij_x = rij_x - box(1) * nint(rij_x * bx)

      rij_y = r(jj, 2) - r(ii, 2)
      rij_y = rij_y - box(2) * nint(rij_y * by)

      rij_z = r(jj, 3) - r(ii, 3)
      rij_z = rij_z - box(3) * nint(rij_z * bz)

      rij = sqrt(rij_x * rij_x + rij_y * rij_y + rij_z * rij_z)
      df = -k(ind) * (rij - r0(ind))

      f(ii, 1) = f(ii, 1) - df * rij_x / rij
      f(jj, 1) = f(jj, 1) + df * rij_x / rij

      f(ii, 2) = f(ii, 2) - df * rij_y / rij
      f(jj, 2) = f(jj, 2) + df * rij_y / rij

      f(ii, 3) = f(ii, 3) - df * rij_z / rij
      f(jj, 3) = f(jj, 3) + df * rij_z / rij

      energy = energy + 0.5d00 * k(ind) * (rij - r0(ind))**2

      bond_pr(1) = bond_pr(1) + ( df * rij_x / rij) * rij_x
      bond_pr(2) = bond_pr(2) + ( df * rij_y / rij) * rij_y
      bond_pr(3) = bond_pr(3) + ( df * rij_z / rij) * rij_z

    end do

end subroutine

File: hymd/compute_bond_forces.c
#include <stdio.h>
#include <math.h>

float cbf(float* f, float* r, float* box, size_t n_bonds, int* a, int* b, float* r0, float* k) {
    size_t aa, bb;
    float df, rab_norm, rab[3], fa[3], energy=0.0;

    for (size_t i = 0; i < n_bonds; i++) {
        aa = 3 * a[i];
        bb = 3 * b[i];

        rab[0] = r[bb] - r[aa];
        rab[1] = r[bb + 1] - r[aa + 1];
        rab[2] = r[bb + 2] - r[aa + 2];

        rab[0] -= box[0] * round(rab[0]/box[0]);
        rab[1] -= box[1] * round(rab[1]/box[1]);
        rab[2] -= box[2] * round(rab[2]/box[2]);

        rab_norm = sqrt(rab[0] * rab[0] + rab[1] * rab[1] + rab[2] * rab[2]);

        df = k[i] * (rab_norm - r0[i]);
        fa[0] = -df * rab[0] / rab_norm;
        fa[1] = -df * rab[1] / rab_norm;
        fa[2] = -df * rab[2] / rab_norm;

        f[aa] -= fa[0];
        f[aa + 1] -= fa[1];
        f[aa + 2] -= fa[2];

        f[bb] += fa[0];
        f[bb + 1] += fa[1];
        f[bb + 2] += fa[2];

        energy += 0.5 * k[i] * pow(rab_norm - r0[i], 2);
    }
    return energy;
}
